---
title: "Class14: Transcriptomics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

For windows, input n when asked to update
```{r}
#source("http://bioconductor.org/biocLite.R")
#biocLite()
```



```{r}
#biocLite("DESeq2")
```

```{r}
library(BiocInstaller)
```


```{r}
counts <- read.csv("Data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("Data/airway_metadata.csv", stringsAsFactors = FALSE)
```


```{r}
head(counts)
```


```{r}
head(metadata)
```

how many genes?
```{r}
nrow(counts)
```

how many colums (experiments)
```{r}
ncol(counts)
```
8 experiments

order of rows in metadata should match counts

```{r}
metadata$id
colnames(counts)[-1]
metadata$id == colnames(counts)[-1]
all(metadata$id == colnames(counts)[-1])
```

Average control experiments and treated experiments and compare


```{r}
metadata[, "dex"]
```


```{r}
metadata[, "dex"] == "control"
```

```{r}
control.ids <- metadata$id[metadata[, "dex"] == "control"]
control.ids

```
```{r}
head(counts)
row.names(counts) <- counts$ensgene
head(counts)
```

```{r}
control <- counts[, control.ids]
control
```


Caclculate mean
```{r}
control.mean <- (rowSums(control))/ncol(control)
head(control)
```


Now for treated
```{r}
metadata[, "dex"] == "treated"
```

```{r}
treated.ids <- metadata$id[metadata[, "dex"] == "treated"]
treated.ids
```

```{r}
treated <- counts[, treated.ids]
treated
```

calculate mean for treated
```{r}
treated.mean <- (rowSums(treated))/ncol(treated)
head(treated)
```


combine control mean and treated mean into a matrix (a column for each)

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```


```{r}
colSums(meancounts)
```

```{r}
plot(meancounts, log = "xy")
```


```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


Filter out genes with zero value

```{r}
#meancounts[, 1:2]
```

```{r}
toy <- c(1, 2, 0, 4, 0)
toy == 0
which(toy == 0)

newtoy <- cbind(toy, toy)
newtoy

newtoy == 0
#just want rows that have 0
unique(which(newtoy == 0, arr.ind = TRUE)[, 1])
```


```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)
head(zero.vals)
```

```{r}
to.rm <- unique(zero.vals[,1])
#all except for those that have zero
mycounts <- meancounts[-to.rm,]
head(mycounts)
```


A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Letâ€™s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

upregulated
```{r}
sum(up.ind)
```

downregulated
```{r}
sum(down.ind)
```

```{r}
head(mycounts[up.ind,])
```

annotation
```{r}
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```

dimensions
```{r}
dim(anno)
```

```{r}
head(mycounts)
```

merge two data frames
what in the first dataframe do you want to combine from the second
```{r}
mycounts.anno <- merge(x = mycounts, y = anno, by.x = "row.names", by.y = "ensgene")
head(mycounts.anno)
```

Install Packages
```{r}
#biocLite("AnnotationDbi")
#biocLite("org.Hs.eg.db")
```


```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```


```{r}
columns(org.Hs.eg.db)
```

Add gene symbol to my counts
```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```

```{r}
head(mycounts)
```


```{r}
library(DESeq2)
citation("DESeq2")
```

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```



```{r}
dds <- DESeq(dds)
```


```{r}
res <- results(dds)
res
```


LFC>0 is up, LFC <0 is down, alpha is p value threshold
```{r}
summary(res, alpha = 0.01)
```


```{r}
plot(res$log2FoldChange, -log(res$padj))
```


```{r}
mycols <- rep("grey", nrow(res))
mycols[abs(res$log2FoldChange) > 2] = "red"
subset.inds <- res$padj < 0.01 & abs(res$log2FoldChange) > 2
mycols[subset.inds] <- "blue"
plot(res$log2FoldChange, -log(res$padj), col = mycols)
```


```{r}
#biocLite("EnhancedVolcano")
```

```{r}

```







